contracts/zero-ex/contracts/src/features/LimitOrdersFeature.sol

```diff
@@ -0,0 +1,875 @@
+            LibLimitOrdersRichErrors.OrderNotFillableByOriginError(
+                orderInfo.orderHash,
+                tx.origin,
+                order.txOrigin
+            ).rrevert();
+        }
+
+        // Signature must be valid for the order.
+        {
+            address signer = LibSignature.getSignerOfHash(orderInfo.orderHash, signature);
+            if (signer != order.maker) {
+                LibLimitOrdersRichErrors.OrderNotSignedByMakerError(
+                    orderInfo.orderHash,
+                    signer,
+                    order.maker
+                ).rrevert();
+            }
+        }
+
+        // Pay the protocol fee.
+        (ethProtocolFeePaid, ) = _collectProtocolFee(order.pool, taker);
+
+        // Settle between the maker and taker.
+        (takerTokenFilledAmount, makerTokenFilledAmount) = _settleOrder(
+            SettleOrderInfo({
+                orderHash: orderInfo.orderHash,
+                maker: order.maker,
+                taker: taker,
+                makerToken: IERC20TokenV06(order.makerToken),
+                takerToken: IERC20TokenV06(order.takerToken),
+                makerAmount: order.makerAmount,
+                takerAmount: order.takerAmount,
+                takerTokenFillAmount: takerTokenFillAmount,
+                takerTokenFilledAmount: orderInfo.takerTokenFilledAmount
+            })
+        );
+
+        emit RfqOrderFilled(
+            orderInfo.orderHash,
+            order.maker,
+            taker,
+            address(order.makerToken),
+            address(order.takerToken),
+            takerTokenFilledAmount,
+            makerTokenFilledAmount,
+            order.pool
+        );
+    }
+
+    /// @dev Settle a the trade between an order's maker and taker.
+    /// @param settleInfo Information needed to execute the settlement.
+    /// @return takerTokenFilledAmount How much taker token was filled.
+    /// @return makerTokenFilledAmount How much maker token was filled.
+    function _settleOrder(SettleOrderInfo memory settleInfo)
+        private
+        returns (uint128 takerTokenFilledAmount, uint128 makerTokenFilledAmount)
+    {
+        // Clamp the taker token fill amount to the fillable amount.
+        takerTokenFilledAmount = LibSafeMathV06.min128(
+            settleInfo.takerTokenFillAmount,
+            settleInfo.takerAmount.safeSub128(settleInfo.takerTokenFilledAmount)
+        );
+        // Compute the maker token amount.
+        // This should never overflow because the values are all clamped to
+        // (2^128-1).
+        makerTokenFilledAmount = uint128(
+            uint256(takerTokenFilledAmount)
+            * uint256(settleInfo.makerAmount)
+            / uint256(settleInfo.takerAmount)
+        );
+
+        if (takerTokenFilledAmount == 0 || makerTokenFilledAmount == 0) {
+            // Nothing to do.
+            return (takerTokenFilledAmount, makerTokenFilledAmount);
+        }
```

Comment:
this can be somewhat misleading, no? We could have `takerTokenFilledAmount > 0` but `makerTokenFilledAmount = 0` due to rounding. I think we should `return(0, 0);` here